src/neo/SmartContract/Native/Nns/NnsContract.cs

```diff
@@ -0,0 +1,310 @@
+using Neo.IO;
+using Neo.IO.Json;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.SmartContract.Native;
+using Neo.SmartContract.Native.Tokens;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Collections;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Text.RegularExpressions;
+using Array = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Nns
+{
+    public partial class NnsContract : Nep11Token<DomainState, Nep11AccountState>
+    {
+        public override int Id => -5;
+        public override string Name => "NNS";
+        public override string Symbol => "nns";
+        public override byte Decimals => 0;
+
+        private const uint BlockPerYear = Blockchain.DecrementInterval;
+        private const uint MaxDomainLevel = 4;
+        private const uint MaxResolveCount = 7;
+
+        private const byte Prefix_Roots = 24;
+
+        internal override bool Initialize(ApplicationEngine engine)
+        {
+            if (!base.Initialize(engine)) return false;
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Roots), new StorageItem(new RootDomainState()
+            {
+                Roots = new Array() { "neo", "wallet", "dapp", "org" }
+            }));
+            return true;
+        }
+
+        [ContractMethod(0_01000000, ContractParameterType.Array, CallFlags.AllowStates)]
+        public StackItem GetRoots(ApplicationEngine engine, Array args)
+        {
+            return engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots)).GetInteroperable<RootDomainState>().Roots;
+        }
+
+        [ContractMethod(1_03000000, ContractParameterType.Boolean, CallFlags.AllowModifyStates, ParameterTypes = new[] { ContractParameterType.ByteArray, ContractParameterType.Hash160 }, ParameterNames = new[] { "name", "owner" })]
+        private StackItem Register(ApplicationEngine engine, Array args)
+        {
+            byte[] tokenId = args[0].GetSpan().ToArray();
+            string name = Encoding.UTF8.GetString(tokenId).ToLower();
+            string[] domains = name.Split(".");
+            if (domains.Length > 3) return false; // only the root and first-level name can be registered
```

Comment:
I think here should be `if (domains.Length > 2) return false;`.

Reply:
?